### DISCUSSION TITLE
What is the reliance on external cloud?

### DISCUSSION THREAD

**Discussion by charlietomo** [sbidy/wiz_light#85] [General]
I am trying to work out what the reliance is on an external cloud for these Wiz bulbs. Ideally I would just use this component to control them locally, with the bulbs restricted from accessing the internet (via my network setup). Has anybody done this or have any insight?

I note that when a bulb is setup, the final stage is "registering on cloud".

**Comment by sbidy** [comment]
I have restricted the bulbs to my network. They can't connect to the cloud service form WiZ. But this will limit the features of the bulb because the app can only communicated via UDP. The UDP API for example doesn't support changing the fade speed and some other "advanced" features of the bulb. But you can change the color, scene, on/off and brightness.

The HASS integration uses this UDP API to communicate lockaly with the bulb. So you can isolate your bulbs to your local network and control them with HASS. I don't use the Cloud API.

Some tech details:
 - The cloud connection is done via MQTT (mqtt-uat.wiz.world, mqtt-dev.wiz.world, us.mqtt.wiz.world)
 - The connections to the cloud are always encrypted with TLS (Lets Encrypt)
 - The bulb will be updated of-the-air via cloud connection (this will done automatically)
 - The Firmware was singed (firmware.wiz.world)
 - The bulb uses a "standard" ESP8266 Hardware
 - The bulb updates the time via ntp.wiz.world

**Reply by charlietomo** [reply]
Thanks that was really helpful.
Do the bulbs need cloud access in order to be setup, or can they be on a restricted network from the start?
Does the Wiz app work at all if the bulbs are restricted from the internet?

**Reply by sbidy** [reply]
Nope - the bulb needs "cloud connection" for the setup. But after it is reachable via Home Wifi you can cut the connection (via FW rule etc.). You can controller the brightness, color and on/off without the cloud. Only "advanced configurations" needs a online bulb. This can be the transition time and custom kelvin range. But if you have configured these setting you can cut the connection again. The settings will be synced to the bulb from the cloud.

**Reply by vodovozovge** [reply]
@charlietomo a bit more details -- you can try to block the WiZ bulb by the mac address on the router and use WiZ App with manual / smart pairing to provision the bulb. Phone needs cloud connection to run through the pairing process.

Keep in mind though that if the bulb cannot connect to the MQTT it will disconnect and connect back to WiFi periodically (happens every several minutes, for ~5 seconds). The idea is to make sure that it's not just a temporary issue with wifi or local network. Other than that -- should work fine. @sbidy could be interesting for you too ;)

WiZ App should also work without internet connection (after the device was provisioned)

**Reply by trimmkm** [reply]
In my limited experience, the Wiz A19 Color bulbs and Home Assistant don't work well when I block access to the Internet.   The bulbs do not respond well to either the Wiz Connected app, or HA.  Looking at my logbook in HA, each of my 4 bulbs "turned off" then "became unavailable" repeatedly every 7 minutes.  I assume this is due to the bulbs disconnecting and reconnecting to wifi as mentioned above.   When I provide them access to the internet, this issue goes away.  The 4 bulbs are in the same group, and it is noticeable when one or more don't respond to commands from HA.  

I recognize this is a feature of the firmware, and beyond our control.  Is there anything I can do to either help lessen the impact to HA when the internet is blocked, or provide info for development/testing?

**Reply by dylan09** [reply]
Thats the exact behaviour I also have seen. My Wiz A19 bulbs regularly disconnect from the AP. 
It stops and they didn't disconnect after I allowed them to connect mqtt.wiz.world (if I remember right) on port MQTT over SSL/TLS (port 8883). They don't need access to any web server or update server. Only to the MQTT server.

I think they do a reboot after they couldn't reach the MQTT server for a few minutes. And thats nothing the integration could help with. Only way would be to reflash them with tasmota or ESPhome.

For meit is acceptable that the bulbs connect to their own MQTT server. They are allowed to connect only one host on port 8883.

**Reply by trimmkm** [reply]
Thanks dylan09 for the info, I'll try to create a URL rule in my router for mqtt.wiz.world.  If it works, this will help limit my exposure.

**Reply by haivala** [reply]
can we replicate mqtt.wiz.world to our own network and make the DNS server point there?

**Reply by dylan09** [reply]
They use MQTT over SSL (port 8883). Maybe its possible to create a DNS entry pointing to a local MQTT broker (eclipse-mosquito). But unfortunately haven't time to test ist. 
But I think without the certificate from Wiz it would not be possible to setup a local server.

I plan to flash the lights to tasmota in the near time future.

**Reply by probonopd** [reply]
Does anyone know full URLs for the OTA on firmware.wiz.world? I would like to have a look at the firmware without taking a device apart. Thanks!

**Reply by rampd2** [reply]
> But I think without the certificate from Wiz it would not be possible to setup a local server.

I think it didn't work as I tried with [eclipse-mosquitto docker image](https://hub.docker.com/_/eclipse-mosquitto/) and as far as I saw, I subscribed to '#' (all topics?), and it didn't output something untill I published some test messages, so I can be sure that my dns setup (with fake answer (eu.)mqtt.wiz.world -> my mqtt broker) seems working.

~If the bulb is just subscribing to the mqtt topic (would explain no messages), **to which topic (1) should I send what messages (2)?**~

Actually I see that it tries to connect over tcp to port 8883 (secure-mqtt) but that does not work, probably because my mqtt broker (mosquitto) does not correctly listen on (just added as additional port host:8883 -> container:1883) and I don't have a valid certificate for `eu.mqtt.wiz.world`.

The broker console displays:
```
New connection from wizbulb-address:port on port 1883.
Client <unknown> disconnected due to malformed packet.
```

**Reply by sbidy** [reply]
So WiZ was acquired by Signify (Philips Lighting) . Maybe there are some changes coming in the future (maybe connect WiZ bulbs to the Hue bridge)

**Reply by GoobyCorp** [reply]
> They use MQTT over SSL (port 8883). Maybe its possible to create a DNS entry pointing to a local MQTT broker (eclipse-mosquito). But unfortunately haven't time to test ist. But I think without the certificate from Wiz it would not be possible to setup a local server.
> 
> I plan to flash the lights to tasmota in the near time future.

I have tested it using no authentication and with self-signed certificates using mosquitto and it does appear like they will connect and accept messages.

![image](https://user-images.githubusercontent.com/4142014/145663902-d1268dbb-e621-4f21-b0b5-dd112cafcae9.png)

**Reply by craigcabrey** [reply]
I was also able to get them to connect to a local MQTT server earlier this year. However I wasn't sure how to figure out what topics they would be listening for. Perhaps if we get a copy of the firmware we can dump strings?

**Comment by GoobyCorp** [comment]
That's certainly possible, I know it just uses an esp8266 like a lot of other smart home devices.

On December 13, 2021 6:17:24 PM EST, Craig Cabrey ***@***.***> wrote:
>I was also able to get them to connect to a local MQTT server earlier this year. However I wasn't sure how to figure out what topics they would be listening for. Perhaps if we get a copy of the firmware we can dump strings?
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/sbidy/wiz_light/discussions/85#discussioncomment-1801980
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by craigcabrey** [comment]
So for what it's worth, it looks like their latest firmware fixes the insecure MQTT (that is, no client side verification). I tested one of my spare bulbs and noticed a message in MQTT that it upgraded its firmware to 1.23:

```
{"method":"devDebug","env":"pro","params":{"rebootReason":"FW 1.23.0(1.22.13)-0x2/1c/402881f5/0/4010093c/38/0/402881f2"}}
```

Now it appears to skip the DNS lookup on bootup and connects directly to Amazon:

```
$ nslookup 18.208.176.54
[...]
Non-authoritative answer:
54.176.208.18.in-addr.arpa	name = ec2-18-208-176-54.compute-1.amazonaws.com.
```

As for how it upgraded firmware, I'm not exactly sure (since I have this VLAN blocked from the internet), but the app may have transmitted an updated FW payload?

**Reply by craigcabrey** [reply]
ahh I take that back, I see it connected now

**Reply by craigcabrey** [reply]
but ultimately, this DEV channel is read only (which makes sense), so we need to get ahold of a firmware dump that's unencrypted to see what methods are available.

it seems to monitor this topic from the logs of the mqtt server

```
1639447638: New client connected from 192.168.20.154:25342 as a8bb50fcce60_3223102 (p1, c1, k40, u'a8bb50fcce60').
1639447638: Will message specified (67 bytes) (r0, q0).
1639447638: 	DEV/pro/a8bb50fcce60
1639447638: Sending CONNACK to a8bb50fcce60_3223102 (0, 0)
1639447638: Received SUBSCRIBE from a8bb50fcce60_3223102
1639447638: 	OP/pro/3223102/devices/a8bb50fcce60 (QoS 0)
1639447638: a8bb50fcce60_3223102 0 OP/pro/3223102/devices/a8bb50fcce60
```

The last part is the MAC address of the bulb, not sure what the 3223102 identifier is. If I publish to that topic, it seems to crash the bulb. Just tried basic stuff like `{"method":"status","params":{}}` -- I'm assuming it's a crash since I see `1639447809: Client a8bb50fcce60_3223102 closed its connection.` immediately after, then it rejoins after 10-15 seconds.

**Reply by GoobyCorp** [reply]
> but ultimately, this DEV channel is read only (which makes sense), so we need to get ahold of a firmware dump that's unencrypted to see what methods are available.
> 
> it seems to monitor this topic from the logs of the mqtt server
> 
> ```
> 1639447638: New client connected from 192.168.20.154:25342 as a8bb50fcce60_3223102 (p1, c1, k40, u'a8bb50fcce60').
> 1639447638: Will message specified (67 bytes) (r0, q0).
> 1639447638: 	DEV/pro/a8bb50fcce60
> 1639447638: Sending CONNACK to a8bb50fcce60_3223102 (0, 0)
> 1639447638: Received SUBSCRIBE from a8bb50fcce60_3223102
> 1639447638: 	OP/pro/3223102/devices/a8bb50fcce60 (QoS 0)
> 1639447638: a8bb50fcce60_3223102 0 OP/pro/3223102/devices/a8bb50fcce60
> ```
> 
> The last part is the MAC address of the bulb, not sure what the 3223102 identifier is. If I publish to that topic, it seems to crash the bulb. Just tried basic stuff like `{"method":"status","params":{}}` -- I'm assuming it's a crash since I see `1639447809: Client a8bb50fcce60_3223102 closed its connection.` immediately after, then it rejoins after 10-15 seconds.

Might be some kind of weird MQTT implementation then, maybe there's some kind of exploit possible to dump the firmware from it.

**Reply by probonopd** [reply]
> dump the firmware from it

Or find out the firmware update URL and download it from there?

**Reply by GoobyCorp** [reply]
> > dump the firmware from it
> 
> Or find out the firmware update URL and download it from there?

Possibly but if it's over HTTPS that probably won't happen.

**Comment by brettonw** [comment]
I use the bulbs on a subnet that does not have internet access. Every so often I open it up and use the app from my phone to connect to the bulbs, and they get firmware updates. That said, they work perfectly with the local only interfaces that sbidy has put together here. It's probably the most reliable component of my home setup, honestly.
